src/top/data/transforms.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python3
+
+import itertools
+from dataclasses import dataclass
+from simple_parsing import Serializable
+from typing import Union, Tuple
+
+import numpy as np
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+
+from torchvision.utils import save_image
+from torchvision.transforms import Compose
+from pytorch3d.transforms.transform3d import Transform3d
+
+from top.run.torch_util import resolve_device
+from top.run.app_util import update_settings
+
+# import cv2
+
+
+class TakeFirst:
+    def __init__(self):
+        pass
+
+    def __call__(self, inputs):
+        return {k: v[0] for (k, v) in inputs.items()}
+
+
+class BoxHeatmap:
+
+    def __init__(self, device: th.device):
+        self.device = device
+        vertices = list(itertools.product(
+            *zip([-0.5, -0.5, -0.5], [0.5, 0.5, 0.5])))
+        vertices = th.as_tensor(vertices, dtype=th.float32, device=self.device)
+        colors = (0.5 + 0.5 * vertices.view(-1, 3, 1, 1)).view(-1, 3)
+        self.vertices = vertices
+        self.colors = colors
+
+    def __call__(self, inputs):
+        # Parse inputs.
+        h, w = inputs['image'].shape[-2:]
+        R = inputs['object/orientation']
+        T = inputs['object/translation']
+        S = inputs['object/scale']
+        P = inputs['camera/projection']
+
+        if 'instance_num' in inputs:
+            num_inst = inputs['instance_num']
+        elif 'num_instances' in inputs:
+            num_inst = inputs['num_instances']
+        else:
+            raise KeyError(F'Instance number not found : {inputs.keys()}')
+
+        heatmaps = []
+        n = int(num_inst)
+        for i in range(n):
+            # NOTE(ycho): Taking an instance-specific slice of batched input.
+            irxn = R[i * 9:(i + 1) * 9]
+            itxn = T[i * 3:(i + 1) * 3]
+            iscale = S[i * 3:(i + 1) * 3]
+
+            # T_scale = np.diag(np.r_[iscale.cpu().numpy(), 1.0])
+            T_scale = Transform3d().scale(iscale.reshape(1,3)).get_matrix()[0].float()
+            print(iscale.shape)
+            print(T_scale.shape)
+            print(T_scale.dtype)
+
+            # BBOX3D transform
+            T_box = th.eye(4)
+            T_box[:3, :3] = irxn.reshape(3, 3)
+            T_box[:3, -1] = itxn
+
+            # camera transforms
+            # T_p = P.cpu().numpy().reshape(4, 4)
+            T_p = P.reshape(4,4).float()#Transform3d(matrix=P.reshape(4,4))
+
+            # Compose all transforms
+            # NOTE(ycho): Looks like `camera/view` is not needed.
+            # Perhaps it's been fused into
+            # object/{translation,orientation}.
+            T = T_p @ T_box @ T_scale
+            print(T.shape) # 3,4,4
+
+            # Apply transform.
+            # NOTE(ycho): skipping division on the last axis here.
+            v = self.vertices @ T[:3, :3].T + T[:3, -1]
+            v[..., :-1] /= v[..., -1:]
+
+            # objectron
```

Comment:
confused comment for me. What about objectron?

Reply:
Yep, fixed.